source/hedgehog/Main.hs

```diff
@@ -0,0 +1,419 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LexicalNegation #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+import qualified Control.Monad.Trans.Writer as Writer
+import qualified Data.Int as Int
+import qualified Data.Set as Set
+import qualified Data.Typeable as Typeable
+import qualified Data.Void as Void
+import qualified Data.Word as Word
+import qualified Hedgehog as H
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Main as Main
+import qualified Hedgehog.Range as Range
+import qualified Numeric.Natural as Natural
+import qualified Witch
+
+main :: IO ()
+main = Main.defaultMain $ fmap H.checkParallel groups
+
+groups :: [H.Group]
+groups = Writer.execWriter $ do
+  Writer.tell $ pure groupWitch
+  Writer.tell $ pure groupInt8
+  Writer.tell $ pure groupInt16
+  Writer.tell $ pure groupInt32
+  Writer.tell $ pure groupInt64
+  Writer.tell $ pure groupInt
+  Writer.tell $ pure groupInteger
+  Writer.tell $ pure groupWord8
+  Writer.tell $ pure groupWord16
+  Writer.tell $ pure groupWord32
+  Writer.tell $ pure groupWord64
+  Writer.tell $ pure groupWord
+  Writer.tell $ pure groupNatural
+  Writer.tell $ pure groupFloat
+  Writer.tell $ pure groupDouble
+
+groupWitch :: H.Group
+groupWitch =
+  H.Group
+    "witch"
+    [ (,) "tripping a a"
+        . H.property
+        . fromFrom @Word.Word8
+        $ Gen.word8 Range.exponentialBounded,
+      (,) "tripping (List a) (Set a)"
+        . H.property
+        . fromFrom @(Set.Set Word.Word8)
+        . Gen.list (Range.linear 0 10)
+        $ Gen.word8 Range.exponentialBounded
+    ]
+
+groupInt8 :: H.Group
+groupInt8 = group "Int8" $ do
+  let gen lo hi = Gen.integral $ Range.linear lo hi :: H.Gen Int.Int8
+  property "Int16" . fromTryFrom @Int.Int16 $ gen -128 127
```

Comment:
It's pretty tedious to come up with these bounds. I'm pretty sure I've gotten them right, but it's hard to check. Perhaps I should write a helper to generate the bounds automatically. Something like this, but shorter:
```suggestion
  property "Int16" . fromTryFrom @Int.Int16 $ gen
    (fromJust . toIntegralSized @Int8 $ max (toInteger $ minBound @Int8) (toInteger $ minBound @Int16))
    (fromJust . toIntegralSized @Int8 $ min (toInteger $ maxBound @Int8) (toInteger $ maxBound @Int16))
```
For unbounded types like `Integer` I would have to pick the bounds arbitrarily, which I'm already doing. And for non-integral types like `Double` I'd also have to pick the bounds arbitrarily, but they're already picked for me (like `maxDouble`).